test/demos/reversediffzero.jl

```diff
@@ -0,0 +1,132 @@
+"The simplest viable reverse mode a AD, only supports `Float64`"
+module ReverseDiffZero
+using ChainRulesCore
+using Test
+
+struct Tracked <: Real
+    propagate::Function
+    primal::Float64
+    tape::Vector{Any}  # a reference to a shared tape
+    grad::Base.RefValue{Float64} # current accumulated gradient
+end
+
+"An intermediate value, a Branch in Nabla terms."
+function Tracked(propagate, primal, tape)
+    v = Tracked(propagate, primal, tape, Ref(zero(primal)))
+    push!(tape, v)
+    return v
+end
+
+"An input, a Leaf in Nabla terms. No inputs of its own to propagate to."
+function Tracked(primal, tape)
+    # don't actually need to put these on the tape, since they don't need to
+    # propagate anything
+    return Tracked(_->nothing, primal, tape, Ref(zero(primal)))
+end
+
+primal(d::Tracked) = d.primal
+primal(d) = d
+
+grad(d::Tracked) = d.grad[]
+grad(d) = nothing
+
+tape(d::Tracked) = d.tape
+tape(d) = nothing
+
+# we have many inputs grab the tape from the first one that is tracked
+get_tape(ds) = something(tape.(ds)...)
+
+# propagate the currently stored gradient back to my inputs.
+propagate!(d::Tracked) = d.propagate(d.grad[])
+propagate!(d) = nothing
+
+# Accumulate gradient, if the value is being tracked.
+accum!(d::Tracked, x̄) = d.grad[] += x̄
+accum!(d, x̄) = nothing
+
+function define_tracked_overload(sig)
+    opT, argTs = Iterators.peel(sig.parameters)
+    fieldcount(opT) == 0 || return  # not handling functors 
+    all(Float64 <: argT for argT in argTs) || return  # only handling purely Float64 ops.
+
+    N = length(sig.parameters) - 1  # skip the op
+    fdef = quote
+        # we use the function call overloading form as it lets us avoid namespacing issues
+        # as we can directly interpolate the function type into to the AST.
+        function (op::$opT)(tracked_args::Vararg{Union{Tracked, Float64}, $N}; kwargs...)
+            args = (op, primal.(tracked_args)...)
+            res = rrule(args...; kwargs...)
+            res === nothing && error("Apparently no rule for $($sig)), but we really thought there was, args=($args)")
+            y, y_pullback = res
```

Comment:
Maybe we should get rid of this line, the code is simpler and easier to read without it,
and we should still get most of the info in the stack-trace if there was an error.
```suggestion
            y, y_pullback = rrule(args...; kwargs...)
```

if we remove it from here then we should also remove it from ForwardDiffZero